### ISSUE TITLE
Confusion between conditions?

### ISSUE THREAD

**Issue by Simo-JDev** [philipperemy/cond_rnn#41]
Hello,

I have a LSTM RNN with the ConditionalRecurrent layer, for a time series prediction where I input the last 10 timesteps and a condition. It seems to me that at times, the conditions get confused, since when testing after training, the outputs of some predictions look extremely similar to the ones expected for a different condition than the one provided. 
More precisely:
Let's say I have 5 possible conditions, A,B,C,D or E (that are float values). While training and checking the predictions every N epochs, I noticed that the model gets really good at predicting when the condition is A, B, C or D but when the condition is E, the output looks similar to what is to be expected if the condition was A, for example. If I train for more epochs, the situation is the same, but for different conditions, i.e. A, B and E work fine, but D predictions look like B predictions.

Is this something that can happen? Otherwise it means that my implementation of the ConditionalRecurrent layer is incorrect.

**Comment by philipperemy** [comment]
The way it works is as follows:

num_units = 10 for the LSTM.

Consider 2 conditions:
- num_1.shape= (batch_size, 1)
- cat_1.shape= (batch_size, 5)

Each condition goes through a different dense layer to output a tensor of shape (batch_size, 10).

We now have [ (batch_size, 10), (batch_size, 10) ]. We stack them on another dimension: (batch_size, 10, 2).

We then apply another Dense layer to project to (batch_size, 10, 1). E.g. we apply a matrix with just 2 weights here 2x1 + 1 bias.

We now have (batch_size, 10, 1). We squeeze the last dimension to (batch_size, 10).

This tensor is the input to the initial states of the recurrent network.

Regarding your example, I'm not sure why you have this behavior. I'd need to think more about it.

**Comment by Simo-JDev** [comment]
I think I did't explain myself very well. The condition is only one, in the plot below it's the `poling_period` layer. It can only assume A, B, C, D or E value, where each value is a float (e.g. 3.7, 3.8, 3.9 etc...).  Does your answer mean that the `cond_rnn` layer needs to go through a Dense layer? I had been referring to the example at [cond_rnn/examples/test_cond_rnn.py](https://github.com/philipperemy/cond_rnn/blob/master/examples/test_cond_rnn.py) since it seemed to be the most applicable to my needs.

![Model Plot](https://user-images.githubusercontent.com/44927443/224025208-8ea447fa-d308-4007-93b0-bea3134c1a72.png)

**Comment by philipperemy** [comment]
You are using it the correct way from what I can see with your plot.

I think you should try to apply a one hot transformation to your conditions since they are discrete:
- 3.7 -> [1, 0, 0, 0, 0]
- 3.8 -> [0, 1, 0, 0, 0]
- [...]

Or if you really want to keep the value you can do:
- 3.7 -> [3.7, 0, 0, 0, 0]
- 3.8 -> [0, 3.8, 0, 0, 0]
- [...]

I'm betting this should solve your problem.

**Comment by Simo-JDev** [comment]
Thanks for your reply, I actually ended up solving the issue by following more closely your example at [cond_rnn/examples/test_cond_rnn.py](https://github.com/philipperemy/cond_rnn/blob/master/examples/test_cond_rnn.py).  The new network looks like this:


![Screenshot 2023-03-13 at 15 23 00](https://user-images.githubusercontent.com/44927443/224730387-1e48efed-fdad-4362-b9b2-4ded0071e56d.png)


It now works great and does not confuse the conditions at all. 
Building on your answer though, I was wondering something. Do you think the ConditionalRecurrent layer/method could be used for a more variable condition? What I mean by that is, if I train the model for the values of the condition of 3.7, 3.8, 3.9, which is now working great, would I then be able to use the NN for a prediction where the condition is set to 3.6 or 3.75? From my testing so far it does not seem to work, but I was wondering if there is a way to make it work.

**Comment by philipperemy** [comment]
@Simo-JDev great your architecture makes sense to me. The condition is fed in both RNN and they should have their states properly initialized to minimize the loss. Regarding your question for the condition, I'd advise you to normalize this condition if possible. Closer to something centered with a variance of 1. Neural Networks are pretty bad at interpolating on large values   (consider 3+ to be a large value). Try something like (X-AVG(X))/STD(X) on your condition.

**Comment by Simo-JDev** [comment]
Sorry for re-opening the issue, but I have a problem linked to this one so I thought it would be helpful to see the old posts. I am trying to make a similar RNN as above, but with two different conditions. One of them is the poling period (as above) so a  value that I normalised as suggested by you. The other is a switch (to choose between the real or imaginary part of the complex signal to predict). 

Initially I tried to replicate the model above, but with the one condition fed to first ConditionalRecurrent, and the other one fed at the second ConditionalRecurrent layer. This is a diagram of the model:
![download](https://github.com/philipperemy/cond_rnn/assets/44927443/05539058-d513-4935-8f04-0daba3277ba7)


This approach does not seem to work really well for the predictions.

I therefore tried to feed each condition twice, as that seemed to help the last time. The model now looks like this:
![download-1](https://github.com/philipperemy/cond_rnn/assets/44927443/48cc303c-56ba-4ec8-8bac-3e53b2a2a144)


I have a problem training this model though. While training, the RAM and Swap Files get filled all the way up to 50 GB after only 5-6 epochs, after which the training stops due to lack of memory. 

My questions are:
1. Is this a problem you have seen before? Is it an issue with keras/tf or with the model?
2. Is there a better way to combine two conditions, considering one is a 'boolean' (although it is currently inputted as 0 or 1) and the other is a normalized value?

Consider that the first model (with one condition fed twice) works well for both conditions individually, I would just like to be able to combine them into one model.
